### DISCUSSION TITLE
prometheus metrics

### DISCUSSION THREAD

**Discussion by Tricked-dev** [dani-garcia/vaultwarden#2952] [Ideas]
being able to make dashboards to visualize Vaultwarden stats would be useful for example list the amount of registered users / total logins

**Comment by BlackDex** [comment]
You could extract some that info via the admin interface and then write a prom file your self. Or maybe extract that info from the database even.
I'm not a fan of adding this to Vaultwarden it self though.

**Reply by Tricked-dev** [reply]
Good idea would have been easier to have a endpoint but this is also a possibility

**Reply by Tricked-dev** [reply]
i have created something simple for anyone wondering <https://github.com/Tricked-dev/vwmetrics>

**Reply by BlackDex** [reply]
@Tricked-dev i have pushed a PR to your repo which will ensure that it opens the database in Read Only mode, to prevent a corrupted database.

Nice little tool btw :).

**Reply by tessus** [reply]
> i have created something simple for anyone wondering https://github.com/Tricked-dev/vwmetrics

Awesome. I have 2 comments though:

- Prometheus metrics services/apps are usually called exporters, thus in this case it would be vaultwarden-exporter (this is not important at all, just something I've noticed)
- a `select count(*)` can be a problem. if there's no index, a table scan has to be performed, which is slow and requires all data to be fetched into the bufferpool, which has other consequences.

But I love it that you wrote this. First time I see this done in Rust. All the other exporters I saw were written in Go.

**Reply by Tricked-dev** [reply]
i like vwmetrics but i'll add 'vaultwarden-exporter' to the topics so people can use it as a search term also what alternative could be used for `select count(*)`?

**Reply by tessus** [reply]
> i like vwmetrics but i'll add 'vaultwarden-exporter' to the topics so people can use it as a search term

good idea. I like it.

> what alternative could be used for select count(*)?

Ha, this depends on a few things, like the database engine, the db structure and/or indexes. I would have top look at the database first.

One way where you can be absolutely sure that no table scan is used is when:
- there's a primary key on the table (which also means there is a unique index on that column)
- you use `select count(pk_column)` in the query

It is possible that a database engine's optimizer uses an index automatically (if there's at least one index defined on a table), when the statement `count(*)` is run. I know that DB2 does that because I was a DB2 developer, but I can't speak for other engines.

**Reply by tessus** [reply]
Most rdbms have a way to show the access plan (explain statement). You can run your statements and check, whether indexes are used. If so, all is good. ;-)

**Reply by BlackDex** [reply]
Vaultwarden doesn't actually uses indexes in the sense that it creates them. There are indexes for the Primary Keys, but that's it as far as i know. Also, `explain` on sqlite is very different compared to MySQL or PostgreSQL in my experience. The numbers and rows it outputs do help a bit, but with joins etc.. it's complex.

**Reply by tessus** [reply]
Yep, I wasn't really thinking of sqlite, since it is not a "real" database engine. IMO it is great for testing and single user workloads, but sqlite should not be used in any scenario where multiple users (processes, threads, connections) have to access the db at the same time.

**Comment by tessus** [comment]
Coming from a database background and having worked in o11y I rather think that using OpenTelemetry for spans and traces would make the devs' lives a lot easier, while Prometheus metrics are more geared towards the user.

I am not saying that Prometheus metrics are bad or not useful. Adding an endpoint e.g. `/admin/metrics` wouldn't hurt performance at all, because the stats would only be available when
- authenticated and
- only generated when accessed/scraped.

A user who doesn't use Prometheus would never hit that endpoint and thus not waste any CPU cycles.

**Reply by BlackDex** [reply]
Traces and spans are in my opinion mostly for in-house applications. You don't want all the logs of all the self-hosted instances going to a central point.

Adding a /metrics endpoint and not using it doesn't hurt indeed. But using it does, depending on the scrape interval. On the server side we could force a 1m or 5m cache to lower the load.
Also, it can contain sensitive info which, if not properly protected could cause issues.

**Reply by tessus** [reply]
Traces and spans can always be generated (the code should be instrumented). It doesn't mean they have to be sent to a backend. The point is that for debugging purposes, especially in distributed systems (or async applications), traces are more useful than log messages. You can always add those messages to spans as event attributes.
Traces are not only useful for debugging and development, but also very useful in production systems. I am not saying one must use them. Having the code properly instrumented is the key. What you do with the data is another story. (This also means you don't have to send/collect the traces at all, but you can, when you want to.)

`/metrics` endpoints provide stats and specific metrics, thus usually do not contain any sensitive/confidential data. This is also underscored by the fact that most endpoints do not use any transport security like TLS.
Even complex stats require in most cases less than 1% of CPU consumption or at least this is what I have seen so far.

Update: But even if you had any sensitive metrics, they'd be behind the admin auth anyway.

**Reply by BlackDex** [reply]
I'm not saying that traces and spans are bad. They could even help a bit in logging indeed in async code.

I have worked for companies where numbers, even user count etc.. could be sensitive. So those metrics could be sensitive. Not saying that i think it will be that big of an issue for Vaultwarden. A default disable could help of course.

**Reply by tessus** [reply]
> I have worked for companies where numbers, even user count etc.. could be sensitive.

I agree. What is considered sensitive cannot be generalized. I'm only saying that the data that exporters generate by default is _usually_ not sensitive.

e.g. when you work for the gov, the number of times the lights go on in a bathroom is considered confidential. ;-)
